### DISCUSSION TITLE
Resuming workflow instances in V3

### DISCUSSION THREAD

**Discussion by gtejeda** [elsa-workflows/elsa-core#6042] [Q&A]
Greetings,
faithful V2 user joining V3 community; I used to resume workflows programmatically doing something like this:

```
string workflowInstanceId = request.WorkflowInstanceId;
var _activity = request.ActivityId;

var _instance = await _workflowInstanceStore.FindByIdAsync(workflowInstanceId);
if (_instance != null)
{
    _instance.Variables.Set("IncomingMsg", request.IncomingMessage);
     
    if (_instance.WorkflowStatus == Elsa.Models.WorkflowStatus.Suspended)
    {
         Console.WriteLine($"Continue workflow with activity: {_activity}");
		 var _result = await _workflowResumer.ResumeWorkflowAsync(_instance, _activity);
		 var workflowStatus = resultado1.WorkflowInstance.WorkflowStatus;
	}	
}
```

Also, suspending the Activity was simply a matter of using return Suspend(); before and the resuming would trigger: ValueTask<IActivityExecutionResult> OnResumeAsync

--------------------------
For v3, I'm able to dispatch workflows using now the elsa's api:

```
POST https://localhost:7043/elsa/api/workflow-definitions/467da8871e09a27e/dispatch HTTP/1.1
Content-Type: application/json
Authorization: ApiKey 00000000-0000-0000-0000-000000000000

{
    "input": {
        "parameter1": "value1",
        "parameter2": "value2"
    }
}
```

however, I can't find an equivalent endpoint for resuming instances... or how to do it programmatically.

I'd greatly appreciate if any of you could point me in the right direction.

Thanks!
Geo

**Comment by jirimatejka** [comment]
I also tried to figure out how to resume workflow via API, but haven't found a way how to do that.

**Reply by stefanie24** [reply]
Me too! Can you access your suspended tasks from Elsa Studio or only from the database?

**Reply by jirimatejka** [reply]
From Elsa studio I can just see the workflow status. 

To resume the workflow I authored my own api, that consumes workflow identifier and bookmark identifier and then executes following code.

var pars = new ResumeWorkflowRuntimeParams { BookmarkId = theBookmarkId };
await wfRuntime.ResumeWorkflowAsync(theWorkflowId, pars);

**Reply by malikhasan** [reply]
This is now marked as obsolete, not sure what the replacement is
"'IWorkflowRuntime.ResumeWorkflowAsync(string, ResumeWorkflowRuntimeParams?)' is obsolete: 'Use the client API instead, retrieved from CreateClientAsync'".

There is one API that uses taksId, but cant find the one that just takes the workflowInstnaceId and bookmarkId

Edit: This works (and there are many variations for it)

 var bookmarkQueueItem = new NewBookmarkQueueItem
 {
     BookmarkId = bookmarkId
 };
 await bookmarkQueue.EnqueueAsync(bookmarkQueueItem);

**Reply by corsjune** [reply]
Not sure if this helps as I am trying to hack this together, but this is how I am resuming using instanceid and bookmarkid. I stoled this from samples or unittests... 

                       //create client using previous ID
                        var workflowClient = await _workflowRuntime.CreateClientAsync(workflowInstance.Id); 

                        var exists = await workflowClient.InstanceExistsAsync();

                        if (!exists)
                            return null;

                       //Start the flow, I will only have the one bookmark
                        var runWorkflowRequest = new RunWorkflowInstanceRequest
                        {
                            Input = new Dictionary<string, object>
                            {
                                { "OptionsVariable", swOptions },
                                { nameof(ContactAction), ca }
                            },
                            BookmarkId = workflowInstance.WorkflowState.Bookmarks.FirstOrDefault()?.Id
                        };

                        var response = await workflowClient.RunInstanceAsync(runWorkflowRequest);
                        
                        //need to get the output property, so lookup hthe instance as it is not exposed on the response
                        workflowInstance = await _workflowInstanceStore.FindAsync(
                            new WorkflowInstanceFilter() { Id = response.WorkflowInstanceId }
                        );

**Comment by sfmskywalker** [comment]
Short answer: the old `IWorkflowRuntime.ResumeWorkflowAsync(...)` path is now obsolete, and the replacement is the newer client/runtime API rather than a generic REST endpoint.

In the current code:

- `IWorkflowRuntime.ResumeWorkflowAsync(string workflowInstanceId, ResumeWorkflowRuntimeParams? options = default)` is explicitly marked obsolete with “Use the client API instead, retrieved from `CreateClientAsync`” (`src/modules/Elsa.Workflows.Runtime/Contracts/IWorkflowRuntime.cs:63-68`).
- `RunWorkflowInstanceRequest` exposes `BookmarkId` directly (`src/modules/Elsa.Workflows.Runtime/Messages/RunWorkflowInstanceRequest.cs:17-20`).
- The repo’s component tests resume bookmarked workflows by creating a client and then calling `RunInstanceAsync` with that bookmark ID (`test/component/Elsa.Workflows.ComponentTests/Scenarios/Variables/CountdownWorkflowTests.cs:22-30` and `48-54`).
- If you want to resume by bookmark/workflow IDs without the client wrapper, `IWorkflowResumer.ResumeAsync(ResumeBookmarkRequest)` is also available (`src/modules/Elsa.Workflows.Runtime/Contracts/IWorkflowResumer.cs:32-38`).

There is also a built-in HTTP endpoint at `/bookmarks/resume`, but it is token-based: it expects a SAS token and then internally constructs a `ResumeBookmarkRequest` from the bookmark/workflow IDs embedded in that token (`src/modules/Elsa.Workflows.Api/Endpoints/Bookmarks/Resume/Endpoint.cs:24-35` and `146-174`).

So the practical answer is:

- for your own code, use `CreateClientAsync(workflowInstanceId)` + `RunInstanceAsync(new RunWorkflowInstanceRequest { BookmarkId = ... })`, or call `IWorkflowResumer` directly;
- for external/public resumption, the stock HTTP surface is the SAS-token `/bookmarks/resume` endpoint, not a plain “instance ID + bookmark ID” Swagger endpoint.
